keps/2724-topology-aware-schedling/README.md

```diff
@@ -0,0 +1,759 @@
+expected to compete for resources with DaemonSet pods.
+
+One way to mitigate the risk is to implement tracking for DaemonSets
+(see [Account usage by watching DaemonSet pods](#account-usage-by-watching-daemonset-pods)).
+We will re-evaluate the need for Beta or GA based on the users' feedback.
+
+## Design Details
+
+### Hierarchy representation
+
+We propose the model for representing the hierarchy of nodes within a data center
+by using node labels. We assume the node labels are set up by a cloud provider,
+or set up manually by administrators of on-premise clusters.
+
+Additionally, we assume that every node used for TAS has a set of the labels
+which identifies uniquely its location in the tree structure. We do not assume
+global uniqueness of labels on each level, i.e. there could be two nodes with
+the same "rack" label, but in different "blocks".
+
+For example, this is a representation of the dataset hierarchy;
+
+|  node  |  cloud.provider.com/topology-block | cloud.provider.com/topology-rack |
+|:------:|:----------------------------------:|:--------------------------------:|
+| node-1 |               block-1              |              rack-1              |
+| node-2 |               block-1              |              rack-2              |
+| node-3 |               block-2              |              rack-1              |
+| node-4 |               block-2              |              rack-3              |
+
+Note that, there is a pair of nodes, node-1 and node-3, with the same value of
+the "cloud.provider.com/topology-rack" label, but in different blocks.
+
+### Admin-facing API
+
+```golang
+// ResourceFlavorSpec defines the desired state of the ResourceFlavor
+type ResourceFlavorSpec struct {
+    ...
+
+  // TopologyName indicates the name of the topology for the ResourceFlavor.
+  // When specified, it enables scraping of the topology information from the
+  // nodes matching to the Resource Flavor node labels.
+  //
+  // +optional
+  TopologyName *string `json:"topologyName,omitempty"`
+}
+
+// TopologySpec defines the desired state of Topology
+type TopologySpec struct {
+  // Levels defines the levels of topology.
+  //
+  // +listType=atomic
+  // +kubebuilder:validation:MinItems=1
+  // +kubebuilder:validation:MaxItems=5
+  Levels []TopologyLevel `json:"levels,omitempty"`
+}
+
+// TopologyLevel defines the desired state of TopologyLevel
+type TopologyLevel struct {
+  // NodeLabel indicates the name of the node label for a specific topology
+  // level. Examples:
+  // - cloud.provider.com/topology-block
+  // - cloud.provider.com/topology-rack
+  //
+  // +required
+  // +kubebuilder:validation:Required
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=316
+  // +kubebuilder:validation:Pattern=`^([a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*/)?(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])$`
+  NodeLabel string `json:"nodeLabel,omitempty"`
+}
```

Comment:
Will we want to support tolerations here based on user feedback?
I agree with supporting only nodeLabel during the alpha stage.

Reply:
I think having different set of tolerations depending on the hierarchy level would be a complication. I don't see it necessary, but yes, if there are use-case we will have a place to introduce it.

Reply:
That makes sense. Thanks for the clarifications.